protocols/request-response/src/lib.rs

```diff
@@ -827,6 +837,15 @@ where
                 self.pending_events
                     .push_back(ToSwarm::GenerateEvent(Event::Message { peer, message }));
             }
+            handler::Event::IncomingRequest { request_id } => {
```

Comment:
I don't understand what the new `Event::IncomingRequest` is needed for. Why does the `ConnectionHandler` need to notify the `NetworkBehaviour` that there is a request coming? Isn't the later on forwarded request itself good enough?

Reply:
I may remove this mechanism. In the current implementation of inbound requests, the task may fail at [`read_request`](https://github.com/libp2p/rust-libp2p/blob/71354b2f42d881b325f5761f5b4748485dcffddb/protocols/request-response/src/handler.rs#L134-L139), which is before user gets notified with any event that contains `InboundRequestId`.

In general, these two scenarios may happen:

1. `Event::Request` -> `Event::InboundFailure` (when `read_request` succeeds, but `write_response` fails).
2. `Event::InboundFailure` (when `read_request` fails).

I thought that the above may look weird to the user of the library, so I introduced the `IncomingRequest` and the above scenarios became like this:

1. `Event::IncomingRequest` -> `Event::Request` -> `Event::InboundFailure`
2. `Event::IncomingRequest` -> `Event::InboundFailure`

I'm still not sure if I like it or not. So let me know what you think.

Reply:
I think I'd rather change the handling of `Event::InboundFailure` to not fail if we have never seen the request before. Ideally, we don't even notify the behaviour of it but I think that would make the error handling code in the handler too complicated :)

Reply:
Additionally, I'd suggest that if we _don't_ know about a request, we also don't notify the user about the `InboundFailure`. There is nothing they can do about the event as they've never seen the ID before.

Reply:
I see two ways of not notifying the upper layer and the user about the failure. But both are "hacky".

1. If `read_request` fails, we return `Ok(())` from the task (we can add a debug log there). No one will see the request ID, since it requires a succeful `read_request` for that.
2. Instead of having a `debug_assert` [here](https://github.com/libp2p/rust-libp2p/blob/71354b2f42d881b325f5761f5b4748485dcffddb/protocols/request-response/src/lib.rs#L928-L936), we can have `if removed { emit_inbound_failure() }`

Do you prefer one of the two?

Reply:
A 3rd way is to keep track the pending request IDs in `Handler` too.

Reply:
I don't consider (2) to be hacky it all. We simply handle the case properly that we might have an `InboundFailure` as the first event.

Reply:
Done